services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py

```diff
@@ -104,7 +126,92 @@ def _replace_env_vars_in_compose_spec(
     return stringified_service_spec
 
 
-def assemble_spec(
+async def _inject_resource_limits_and_reservations(
+    app: FastAPI, service_key: str, service_tag: str, service_spec: ComposeSpecLabel
+) -> None:
+    director_v0_client: DirectorV0Client = get_director_v0_client(app)
+    labels_for_involved_services: Dict[
+        str, SimcoreServiceLabels
+    ] = await get_labels_for_involved_services(
+        director_v0_client=director_v0_client,
+        service_key=service_key,
+        service_tag=service_tag,
+    )
+
+    # example: '2.3' -> 2 ; '3.7' -> 3
+    docker_compose_major_version: int = int(service_spec["version"].split(".")[0])
+
+    for spec_service_key, spec in service_spec["services"].items():
+        if spec_service_key not in labels_for_involved_services:
+            logger.info(
+                "No labels found for service %s, service may not be an osparc service",
+                spec_service_key,
+            )
+            continue
+        labels = labels_for_involved_services[spec_service_key]
+
+        settings_list: List[SimcoreServiceSettingLabelEntry] = labels.settings
+
+        # defaults
+        limit_nano_cpus = DEFAULT_LIMIT_NANO_CPUS
+        limit_memory_bytes = DEFAULT_LIMIT_MEMORY_BYTES
+        reservation_nano_cpus = DEFAULT_RESERVATION_NANO_CPUS
+        reservation_memory_bytes = DEFAULT_RESERVATION_MEMORY_BYTES
+
+        for entry in settings_list:
+            if entry.name == "Resources" and entry.setting_type == "Resources":
+                values: Dict[str, Any] = entry.value
+
+                # fetch limits
+                limits: Dict[str, Any] = values.get("Limits", {})
+                limit_nano_cpus = limits.get("NanoCPUs", DEFAULT_LIMIT_NANO_CPUS)
+                limit_memory_bytes = limits.get(
+                    "MemoryBytes", DEFAULT_LIMIT_MEMORY_BYTES
+                )
+
+                # fetch reservations
+                reservations: Dict[str, Any] = values.get("Reservations", {})
+                reservation_nano_cpus = reservations.get(
+                    "NanoCPUs", DEFAULT_RESERVATION_NANO_CPUS
+                )
+
+                reservation_memory_bytes = reservations.get(
+                    "MemoryBytes", DEFAULT_RESERVATION_MEMORY_BYTES
+                )
+
+                break
```

Comment:
so you expect only one match, right?

why not just searching for it in one go? 

```python

try:
   entry = next( e for e in settings_list if e.name == "Resources" and e.setting_type == "Resources" )
   # ...
except StopIteration:
   pass
```

Reply:
outdated